### PR TITLE
Moved sct_download_data functions to new download module

### PR DESCRIPTION
**Pull request by jcohenadad**
While working on https://github.com/neuropoly/spinalcordtoolbox/pull/2639, I had to make `sct_download_data`'s subfunctions callable by other Python processes, so the purpose of this PR was to move those functions (previously located in the CLI script) into a new `spinalcordtoolbox.download` module. 

This PR also introduces the following changes to `sct.utils`:
- added: `add_suffix` and `splitext` (formerly `extract_fname`)
- created unit tests for these

### CODE REVIEW

File: scripts/sct_download_data.py [hunk 1]
```diff
@@ -51,6 +43,7 @@ def get_parser():
             'optic_models',
             'pmj_models',
             'binaries_debian',
+            'binaries_centos',
```
Comment (kousu) [inline scripts/sct_download_data.py:46]: This was gone in #2699.

File: scripts/sct_download_data.py [hunk 2]
```diff
@@ -38,6 +29,7 @@ def get_parser():
         type_value="multiple_choice",
         description="Name of the dataset.",
         mandatory=True,
+        # TODO: replace with key dict items
```
Comment (kousu) [inline scripts/sct_download_data.py:32]: Are you thinking of loading the docstring directly from `DICT_URL`? `example=sorted(DICT_URL)`?

File: spinalcordtoolbox/utils.py [hunk 3]
```diff
@@ -138,83 +156,19 @@ def is_exe(fpath):
...
-    sct_branch = os.getenv(branch_env, "unknown")
-    if check_exe("git") and os.path.isdir(os.path.join(__sct_dir__, ".git")):
-        sct_commit = __get_commit() or sct_commit
-        sct_branch = __get_branch() or sct_branch
-
-    if sct_commit is not 'unknown':
-        install_type = 'git'
+    parent, filename = os.path.split(fpath)
+    if filename.endswith(".nii.gz"):
+        stem, ext = filename[:-7], ".nii.gz"
```
Comment (kousu) [inline spinalcordtoolbox/utils.py:168]: You could add the same special-case for `.tar.gz`. The number of special cases in here might fast get unwieldy.

File: spinalcordtoolbox/utils.py [hunk 4]
```diff
@@ -138,83 +156,19 @@ def is_exe(fpath):
     return None
 
 
-def __get_branch():
+def extract_fname(fpath):
```
Comment (kousu) [inline spinalcordtoolbox/utils.py:159]: I would just make this

```suggestion
def splitext(filename):
```

If users need the folder path, they already have `os.path.split()`. `sct.utils.splitext()` can be a drop in neuroimaging-aware replacement for `os.path.splitext()`.

File: spinalcordtoolbox/utils.py [hunk 5]
```diff
@@ -115,6 +116,23 @@ def _split_lines(self, text, width):
...
+    :param suffix: suffix. Example: _mean
+    :return: file name with suffix. Example: t2_mean.nii
+
+    Examples:
+
+    - add_suffix(t2.nii, _mean) -> t2_mean.nii
+    - add_suffix(t2.nii.gz, a) -> t2a.nii.gz
+    """
+    parent, stem, ext = extract_fname(fname)
+    return os.path.join(parent, stem + suffix + ext)
```
Comment (kousu) [inline spinalcordtoolbox/utils.py:133]: You don't need to separate the parent:

```
>>> base, ext = os.path.splitext("path/to/dataset/t2.nii.gz")
>>> base + '_mean' + ext
'path/to/dataset/t2.nii_mean.gz'
```

If that was `sct.utils.splitext()` instead that was `.nii.gz`-aware that would have produced the right result.

File: spinalcordtoolbox/utils.py (hunk 3 above)
Comment (kousu) [reply spinalcordtoolbox/utils.py:168]: It almost looks like this exists in the stdlib @ https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffixes but it only does half the job. :abacus:

File: spinalcordtoolbox/utils.py (hunk 5 above)
Comment (kousu) [reply spinalcordtoolbox/utils.py:133]: It handles `.`s in the parent dirs properly:

```
>>> base, ext = os.path.splitext("path/to.tar.gz/dataset/t2.tar.gz")
>>> base + "_mean" + ext
'path/to.tar.gz/dataset/t2.tar_mean.gz'
>>> 
```

File: spinalcordtoolbox/utils.py (hunk 3 above)
Comment (kousu) [reply spinalcordtoolbox/utils.py:168]: What about

```suggestion
    dir, filename = os.path.split(filename)
    stem, ext = filename.split(".",1)
    ext = "." + ext
    return os.path.join(dir, stem), ext
```

File: scripts/sct_download_data.py (hunk 1 above)
Comment (jcohenadad) [reply scripts/sct_download_data.py:46]: fixed in 190acb6bb392c179a7e40e2ea7e5e6f58cfd10b1

File: scripts/sct_download_data.py (hunk 2 above)
Comment (jcohenadad) [reply scripts/sct_download_data.py:32]: yup! fixed in 9d531465638a9b7299f52526eb7b2ba36b0cc5b3

File: spinalcordtoolbox/utils.py (hunk 3 above)
Comment (jcohenadad) [reply spinalcordtoolbox/utils.py:168]: the problem with that implementation is that it fails the following test:
```python
    assert utils.splitext('nice.image.nii.gz') == ('nice.image', '.nii.gz')
```
I've seen such file names in my life 🙄...

Comment (jcohenadad) [reply spinalcordtoolbox/utils.py]: now accommodating tar.gz (and we can add to the list of extensions if we encounter more): f537788664a3fb81876235d67f92ef910f1bd6f6

File: spinalcordtoolbox/utils.py (hunk 4 above)
Comment (jcohenadad) [reply spinalcordtoolbox/utils.py]: great suggestion. fixed in ea0c10616049b8a433c2e7ebba5aedf3e498568b

File: spinalcordtoolbox/utils.py (hunk 3 above)
Comment (kousu) [reply spinalcordtoolbox/utils.py]: > the problem with that implementation is that it fails the following test:
> 
> ```python
>     assert utils.splitext('nice.image.nii.gz') == ('nice.image', '.nii.gz')
> ```
> 
> I've seen such file names in my life roll_eyes...

I was hoping they wouldn't be an issue. Alright, special-case away!

File: scripts/sct_download_data.py [hunk 6]
```diff
@@ -12,23 +12,51 @@
...
+                         'https://www.neuro.polymtl.ca/_media/downloads/sct/20190802_sct_course-beijing19.zip'],
+    'deepseg_gm_models': ['https://osf.io/b9y4x/?action=download',
+                          'https://www.neuro.polymtl.ca/_media/downloads/sct/20180205_deepseg_gm_models.zip'],
+    'deepseg_sc_models': ['https://osf.io/avf97/?action=download',
+                          'https://www.neuro.polymtl.ca/_media/downloads/sct/20180610_deepseg_sc_models.zip'],
+    'deepseg_lesion_models': ['https://osf.io/eg7v9/?action=download',
+                              'https://www.neuro.polymtl.ca/_media/downloads/sct/20180613_deepseg_lesion_models.zip'],
+    'c2c3_disc_models': ['https://osf.io/t97ap/?action=download',
+                         'https://www.neuro.polymtl.ca/_media/downloads/sct/20190117_c2c3_disc_models.zip']
+    }
```
Comment (kousu) [inline scripts/sct_download_data.py:58]: DICT_URL is still getting deleted/added, so it's going to end up make two copies of it when I rebase my PRs. Can you put it back down below `get_parser`?

Changing the name to uppercase should be okay.

File: unit_testing/test_utils.py [hunk 7]
```diff
@@ -7,6 +7,20 @@
 from spinalcordtoolbox import utils
 
 
+def test_add_suffix():
+    assert utils.add_suffix('t2.nii', '_mean') == 't2_mean.nii'
+    assert utils.add_suffix('t2.nii.gz', 'a') == 't2a.nii.gz'
```
Comment (kousu) [inline unit_testing/test_utils.py:12]:
```suggestion
    assert utils.add_suffix('t2.nii.gz', 'a') == 't2a.nii.gz'
    assert utils.add_suffiix('var/lib/usr/t2.nii.gz', 'sfx') == 'var/lib/usr/t2sfx'.nii.gz'
    assert utils.add_suffiix('var/lib.version.3/usr/t2.nii.gz', 'sfx') == 'var/lib.version.3/usr/t2sfx'.nii.gz'
```

File: scripts/sct_download_data.py (hunk 6 above)
Comment (jcohenadad) [reply scripts/sct_download_data.py:58]: i can, but in that case i don't know how to implement [this](https://github.com/neuropoly/spinalcordtoolbox/pull/2700#pullrequestreview-408886355)

Comment (kousu) [reply scripts/sct_download_data.py:58]: That's inside of `get_parser()` so it should be fine -- it doesn't actually run until `main()` calls it.

Comment (kousu) [reply scripts/sct_download_data.py:58]: I know, that list really should be more prominent in this file. But we can do it in a small PR next week.

Comment (jcohenadad) [reply scripts/sct_download_data.py]: hum, i am still getting the error (in 9840ff76f611518c288e3db8c5fdd2dce779e636):
```bash
 sct_download_data -h

--
Spinal Cord Toolbox (git-jca/download-data-9840ff76f611518c288e3db8c5fdd2dce779e636)

Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_download_data.py", line 113, in <module>
    res = main()
  File "/Users/julien/code/sct/scripts/sct_download_data.py", line 97, in main
    parser = get_parser()
  File "/Users/julien/code/sct/scripts/sct_download_data.py", line 32, in get_parser
    example=sorted(dict_url))
NameError: name 'dict_url' is not defined
```

Comment (kousu) [reply scripts/sct_download_data.py]: Oh it was a local variable. I should have tried it myself, sorry.

Two fixes:

1. Very quickly: `parser = get_parser()` -> `parser = get_parser(dict_url)`
2. You could probably move ` def main(args=None): ` *after* `dict_url`; then there will be a small patch on the `dict_url` line but the rest of its contents should stay put.

We'll clean this up better once we've merged our work fully!

File: unit_testing/test_utils.py (hunk 7 above)
Comment (jcohenadad) [reply unit_testing/test_utils.py:12]: done in 58e1d86e8e4c7cb1961a6e68b643947f7390854a

File: scripts/sct_download_data.py (hunk 6 above)
Comment (jcohenadad) [reply scripts/sct_download_data.py]: 👍 i did the first suggestion 8ad4aee81c0a03b42f87cd5a03fb0eade07bf51c

**Pull request merged**